### PR TITLE
for triple-panics, do not keep printing backtraces

### PR DESCRIPTION
**Pull request by RalfJung**
Cc https://github.com/rust-lang/rust/issues/114954
r? @Amanieu

### CODE REVIEW

**Comment by matthiaskrgr** [comment]
Will this change the behavior of `-Ztreat-err-as-bug=x` ?

**Comment by RalfJung** [comment]
I... don't think so? This is a library patch, that's a compiler flag. How does that flag even relate to triple-panics?

**Comment by matthiaskrgr** [comment]
It can easily cause N-panics. I am not too familiar with this unfortunately, but we used this in https://github.com/rust-lang/rust/issues/113211 to find panics in nonuwind functions.

**Comment by RalfJung** [comment]
For such cases you can always set RUST_BACKTRACE to get a backtrace as usual. This only changes whether a backtrace is printed when that variable is not set.

File: library/std/src/panicking.rs [hunk 1]
```diff
@@ -246,7 +246,10 @@ fn default_hook(info: &PanicInfo<'_>) {
 pub fn panic_hook_with_disk_dump(info: &PanicInfo<'_>, path: Option<&crate::path::Path>) {
     // If this is a double panic, make sure that we print a backtrace
     // for this panic. Otherwise only print it if logging is enabled.
-    let backtrace = if panic_count::get_count() >= 2 {
+    // We do not keep printing for further panics, so that e.g. a "panic inside a drop on an unwind
+    // path", which leads to a triple-panic (the third panic being "cannot unwind here"), doesn't
+    // print *two* backtraces.
+    let backtrace = if panic_count::get_count() == 2 {
```
Comment (bjorn3) [inline library/std/src/panicking.rs:252]: For panic count > 2 should this return `None`?

Comment (RalfJung) [reply library/std/src/panicking.rs:252]: Should what return None? The function we are in does not return anything.

Comment (bjorn3) [reply library/std/src/panicking.rs:252]: I meant the `backtrace` variable should be assigned `None` if the panic count > 2. This seems to indicate that backtrace printing is not available.

Comment (RalfJung) [reply library/std/src/panicking.rs:252]: But why would we want that? I think we still want RUST_BACKTRACE to be respected.

Comment (bjorn3) [reply library/std/src/panicking.rs:252]: For a triple panic I can imagine there is a fair chance that the panic actually happens while printing the backtrace and thus we shouldn't try to print a backtrace for the third panic.

Comment (RalfJung) [reply library/std/src/panicking.rs:252]: We have a triple-panic already when just causing a panic during cleanup. So I don't think that this is a good indication for panicking during backtrace printing.

**Comment by RalfJung** [comment]
> I don't think that using the panic count is the right approach here. 

Why not? If the user didn't ask for a backtrace, then printing *two* (or more) of them is surely not what they were looking for.

>  I would prefer if we specifically disabled backtrace printing when the reason is known to be one where we don't need the additional backtrace. 

Are you saying that `PanicInfo` should carry another boolean flag that can be used to disable the backtrace?

**Comment by Amanieu** [comment]
> Why not? If the user didn't ask for a backtrace, then printing _two_ (or more) of them is surely not what they were looking for.

It could be a perfectly valid case of nested panics (with `catch_unwind` called from within `Drop`), where we do want to make sure each panic gets a backtrace printed.



> Are you saying that `PanicInfo` should carry another boolean flag that can be used to disable the backtrace?

That's the best solution I can think of at the moment. At least it's more explicit than just guessing based on the panic count.

**Comment by RalfJung** [comment]
All right. Superseded by https://github.com/rust-lang/rust/pull/115280.

**Pull request closed without merging**